noisy-simulator/src/density_matrix_simulator/tests.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use super::DensityMatrixSimulator;
+use crate::{
+    instrument::Instrument,
+    operation::{operation, Operation},
+    Error, TOLERANCE,
+};
+use num_complex::Complex;
+
+macro_rules! assert_approx_eq {
+    ($left:expr, $right:expr $(,)?) => {
+        if !approx_eq($left, $right) {
+            panic!("aprox_equal failed, left = {}, right = {}", $left, $right);
+        }
+    };
+}
+
+fn approx_eq(a: f64, b: f64) -> bool {
+    (a - b).abs() <= TOLERANCE
+}
+
+fn h() -> Operation {
+    let f = 0.5_f64.sqrt();
+    operation!([f,  f;
+                f, -f;])
+}
+
+fn cnot() -> Operation {
+    operation!([1., 0., 0., 0.;
+                0., 1., 0., 0.;
+                0., 0., 0., 1.;
+                0., 0., 1., 0.;])
+}
+
+fn mz() -> Instrument {
+    let m0 = operation!([1., 0.;
+                         0., 0.;]);
+    let m1 = operation!([0., 0.;
+                         0., 1.;]);
+    Instrument::new(vec![m0, m1])
+}
+
+#[test]
+fn constructor() {
+    let _sim = DensityMatrixSimulator::new(1);
+}
+
+#[test]
+fn one_qubit() {
+    let h = h();
+    let mz = mz();
+
+    for _ in 0..10 {
+        let mut sim = DensityMatrixSimulator::new(1);
+        sim.apply_operation(&h, &[0])
+            .expect("operation should succeed");
+        let measurement = sim
+            .sample_instrument_with_distribution(&mz, &[0], 0.3)
+            .expect("measurement should succeed");
+        assert_eq!(measurement, 0);
+    }
+
+    for _ in 0..10 {
+        let mut sim = DensityMatrixSimulator::new(1);
+        sim.apply_operation(&h, &[0])
+            .expect("operation should succeed");
+        let measurement = sim
+            .sample_instrument_with_distribution(&mz, &[0], 0.7)
+            .expect("measurement should succeed");
+        assert_eq!(measurement, 1);
+    }
+}
+
+#[test]
+fn bell_pair_sampling() {
```

Comment:
If I see a test failure with this name, how does it help me know what was expected, what failed, in what context, so that I can narrow my search to find the bug?

Reply:
Documented this test.